### ISSUE TITLE
Could not connect (sequence) - ESOCKET

### ISSUE THREAD

**Issue by fedecaccia** [tediousjs/tedious#1168]
Hi, Im experimenting an error using mssql library for nodejs. I am using PoolConnection from mssql (and I also tried using tedious-pool-connection library). The app Im running simply connects to a SQL server db and executes some writing queries. Multiple queries at the same time, or in serial order, it doesn't matter: I always get this kind of errors for some queries:

```
ConnectionError: Failed to connect to <DB_URL>:1433 - Could not connect (sequence)
    at Connection.<anonymous> (<PROJECT_DIR>/node_modules/mssql/lib/tedious/connection-pool.js:68:17)
    at Object.onceWrapper (events.js:421:26)
    at Connection.emit (events.js:314:20)
    at Connection.socketError (<PROJECT_DIR>/node_modules/mssql/node_modules/tedious/lib/connection.js:1290:12)
    at <PROJECT_DIR>/node_modules/mssql/node_modules/tedious/lib/connection.js:1116:21
    at SequentialConnectionStrategy.connect (<PROJECT_DIR>/node_modules/mssql/node_modules/tedious/lib/connector.js:87:14)
    at Socket.onError <PROJECT_DIR>/node_modules/mssql/node_modules/tedious/lib/connector.js:100:12)
    at Socket.emit (events.js:314:20)
    at emitErrorNT (internal/streams/destroy.js:92:8)
    at emitErrorAndCloseNT (internal/streams/destroy.js:60:3) {
  code: 'ESOCKET',
  originalError: ConnectionError: Failed to connect to <DB_URL>:1433 - Could not connect (sequence)
      at ConnectionError (<PROJECT_DIR>/node_modules/mssql/node_modules/tedious/lib/errors.js:13:12)
      at Connection.socketError (<PROJECT_DIR>/node_modules/mssql/node_modules/tedious/lib/connection.js:1290:56)
      at <PROJECT_DIR>/node_modules/mssql/node_modules/tedious/lib/connection.js:1116:21
      at SequentialConnectionStrategy.connect (/<PROJECT_DIR>/node_modules/mssql/node_modules/tedious/lib/connector.js:87:14)
      at Socket.onError (<PROJECT_DIR>/node_modules/mssql/node_modules/tedious/lib/connector.js:100:12)
      at Socket.emit (events.js:314:20)
      at emitErrorNT (internal/streams/destroy.js:92:8)
      at emitErrorAndCloseNT (internal/streams/destroy.js:60:3)
      at processTicksAndRejections (internal/process/task_queues.js:84:21) {
    code: 'ESOCKET'
  }
}

```

What could be causing this issue?

**Comment by fedecaccia** [comment]
The dependencies Im using:
test 1: "mssql": "^6.2.1"
and in test 2: "tedious": "^1.14.0", "tedious-connection-pool": "^1.0.5"

The DB instance has not an important load. Im checking all metrics in AWS (an I was also checking all the metrics with AWS support, they agreed everything is heathy from DB perspective and that the issue is related to the app instance).

**Comment by IanChokS** [comment]
Hi @fedecaccia Check out our [FAQ page here](https://tediousjs.github.io/tedious/frequently-encountered-problems.html)

Also tedious 1.14.0 is years old. The latest version is 9.2.1

**Comment by IanChokS** [comment]
Closing this as it is unlikely an issue with the latest version of the driver.

**Comment by fedecaccia** [comment]
> 
> 
> Hi @fedecaccia Check out our [FAQ page here](https://tediousjs.github.io/tedious/frequently-encountered-problems.html)
> 
> Also tedious 1.14.0 is years old. The latest version is 9.2.1

Hi @IanChokS I was using the last version, until I tested using tedious-connection-pool, and it was suggested to me to use this version in order to make it work

**Comment by fedecaccia** [comment]
@IanChokS please re open it, as I described above, I confirm I am using the last version of mssql, and Im also using tediuos 9.2.1.
The only reason I used "tedious": "^1.14.0" is because I was using  "tedious-connection-pool": "^1.0.5" in the second test

**Comment by IanChokS** [comment]
Unfortunately tedious-connection-pool isn't actively maintained. We do[ have a plan ](https://github.com/tediousjs/tedious/milestone/10) to integrate that library into `tedious` itself one of these days. 

Your failure to connect can be of a variety of reasons. What is the config you're passing into `tedious`? Did you ensure your SQL DB is open for connection (you can check in the FAQ I linked above). Is your DB instance hosted on AWS? I believe there have been issues connecting to AWS hosted DB's in the past. Also have you tried connecting using just `tedious`?

**Comment by fedecaccia** [comment]
Ok, tedious-connection-pool  was just an alternative I tried, lets forget about it.
Regarding the SQL DB, I conffirm it is open for connection. We already check with AWS RDS support all kind of configs: regarding security groups, vpc, etc. Everything is fine. The connection is working and most of the queries are executed succesfully.

The configs I pass to tedious:

```
    config = {
        server: <URL_SERVER>,
        database: <DATABASE_NAME>,
        user: <USER_NAME>,
        password: <SQL_PASSWORD>,
        "connectionTimeout": 300000,
        "requestTimeout": 300000,
        "pool": {
            "max": 300,
            "min": 0,
            "idleTimeoutMillis": 30000
        },
        "options": {
            "encrypt": true,
            "enableArithAbort": true
        },
    }
```

**Comment by fedecaccia** [comment]
I create the pool using:

```
function getConnection () {    
    return new Promise(function (resolve, reject) {
        const pool = new sql.ConnectionPool(config)     
        pool.connect()
        .then((cn) => ...)
        .catch((err) => ...)
    });
}
```
Most of the connections are created right, but sometimes executing connect() rejects with the error I submitted previously.

**Comment by IanChokS** [comment]
The config doesn't look right. It should have the form:
```js
var Connection = require('tedious').Connection;
var config = {
  "authentication": {
    ...,
    "options": {...}
  },

  "options": {
    ...
  }
};
var connection = new Connection(config);
```
Note: the `userName` and `password` are in the `authentication.options` and not the `options`, as[ specified in the docs](https://tediousjs.github.io/tedious/api-connection.html)

I apologize for that config confusion, it happens quite a bit which is why it's in the FEQ 😅 

Try using the config in the correct format, hopefully you'll be able to connect then.

**Comment by fedecaccia** [comment]
Thanks, I understand this is not the format expected using just tedious, buy Im using mssql above tedious, and the config input agree with what mssql expects. Im able to connect, and it works propertly most of the time. The ESOCKET error is shooted when Im running loops of queries. So I think the issue is not related with the connection config input, but with some internal management ni the connection pool. 

I detected the error is shoot by the method socketError in tedious/lib/connection.js, which is shot because  this.state === this.STATE.SENT_TLSSSLNEGOTIATION.

**Comment by IanChokS** [comment]
Ah, sorry I thought you were having trouble connecting to your db. 

I'm not familiar with how `mssql` handle's tedious, but for tedious itself you need to wait for a request to finish before making another request (in the callback of the first request) on the same connection. 

For e.g., 
```js
var connection = new Connection(config);

connection.connect(function(err) {
    // If no error, then good to go...
    executeStatement();
  }
);

function executeStatement() {
  request = new Request("select 42, 'hello world'", function(err, rowCount) {
    if (err) {
      console.log(err);
    }

    const request = new Request('`select * from table`, (err, rowCount) => {
        if (err) {
             console.log(err);
        }
           // connection.close() or new Request etc....
     });

    connection.execSql(request);
  });

  connection.execSql(request);
}
```

**Comment by fedecaccia** [comment]
Ok I understand, but you are using simple connections in this example, not a connection pool. In a connection pool there should be a set of available threads that can be used in parallel.

**Comment by IanChokS** [comment]
True, but as I stated before `tedious` doesn't support connection pooling yet. We hope to integrate that library so it can become available. I'm not sure if `mssql` supports connection pooling either, but you can try asking them on[ their repo](https://github.com/tediousjs/node-mssql) 

Also just to confirm, if you run the queries one at a time, do you run into any issues (without any connection pooling)?

**Comment by fedecaccia** [comment]
Ok, yes I referred to connectionPool from mssql so Ill wait for their response on their repo. I haven't tried using simple connection from tedious, Ill also try and let you know. 

One last question, do you know what throws the state SENT_TLSSSLNEGOTIATION?

Thanks for all your help!

**Comment by IanChokS** [comment]
No worries, happy to help 😄 

The `SENT_TLSSSLNEGOTIATION` is the state of tedious when it's trying to establish a connection to the server (in the pre-login stage). This can be thrown in various ways, but I'm guessing you're trying to send a `request` while the driver is not yet connected to the server.

**Comment by fedecaccia** [comment]
Ok, I dont understand which is the reason for that, it all depends on how mssql library is hadling the connection internally. Will check with them. Thanks!

**Comment by fedecaccia** [comment]
I link with the issue in mssql-nodejs: https://github.com/tediousjs/node-mssql/issues/1121

**Comment by IanChokS** [comment]
I'm going to close this for now as it doesn't appear to be an issue when used as it is intended to. If you discover an error from a `tedious` specific use, feel free to comment on here or raise a new issue :)

**Comment by fedecaccia** [comment]
> 
> 
> I'm going to close this for now as it doesn't appear to be an issue when used as it is intended to. If you discover an error from a `tedious` specific use, feel free to comment on here or raise a new issue :)

ok! thanks anyway
